src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/DefaultConverter.cs

```diff
@@ -0,0 +1,64 @@
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.ConvertLinq.ConvertForEachToLinqQuery;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq.ConvertForEachToLinqQuery
+{
+    internal sealed class DefaultConverter : AbstractConverter
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+        public DefaultConverter(ForEachInfo<ForEachStatementSyntax, StatementSyntax> forEachInfo)
+            : base(forEachInfo) { }
+
+        public override void Convert(SyntaxEditor editor, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // Filter out identifiers which are not used in statements.
+            var symbolNames = new HashSet<string>(ForEachInfo.Statements.SelectMany(statement => semanticModel.AnalyzeDataFlow(statement).ReadInside).Select(symbol => symbol.Name));
+            var identifiersUsedInStatements = ForEachInfo.Identifiers.Where(identifier => symbolNames.Contains(identifier.ValueText));
+
+            // If there is a single statement and it is a block, leave it as is.
+            // Otherwise, wrap with a block.
+            var block = WrapWithBlockIfNecessary(ForEachInfo.Statements.Select(statement => statement.KeepCommentsAndAddElasticMarkers()));
+
+            editor.ReplaceNode(ForEachInfo.ForEachStatement, CreateDefaultReplacementStatement(ForEachInfo, identifiersUsedInStatements, block).WithAdditionalAnnotations(Formatter.Annotation));
+        }
+
+        private StatementSyntax CreateDefaultReplacementStatement(
+            ForEachInfo<ForEachStatementSyntax, StatementSyntax> forEachInfo,
+            IEnumerable<SyntaxToken> identifiers,
+            BlockSyntax block)
+        {
+            var identifiersCount = identifiers.Count();
+            if (identifiersCount == 0)
+            {
+                // Generate foreach(var _ ... select new {})
+                return SyntaxFactory.ForEachStatement(VarNameIdentifier, SyntaxFactory.Identifier("_"), CreateQueryExpression(SyntaxFactory.AnonymousObjectCreationExpression(), Enumerable.Empty<SyntaxToken>(), Enumerable.Empty<SyntaxToken>()), block);
+            }
+            else if (identifiersCount == 1)
+            {
+                // Generate foreach(var singleIdentifier from ... select singleIdentifier)
+                return SyntaxFactory.ForEachStatement(VarNameIdentifier, identifiers.Single(), CreateQueryExpression(SyntaxFactory.IdentifierName(identifiers.Single()), Enumerable.Empty<SyntaxToken>(), Enumerable.Empty<SyntaxToken>()), block);
+            }
+            else
+            {
+                var tupleForSelectExpression = SyntaxFactory.TupleExpression(SyntaxFactory.SeparatedList(identifiers.Select(identifier => SyntaxFactory.Argument(SyntaxFactory.IdentifierName(identifier)))));
```

Comment:
I guess identifier is already in correctly escaped format? and no scope issue?

Reply:
1. identifiers are filtered
2. we bail out in case if identifiers have conflicts with the scope
3. I added a unit test to cover escape symbols

thank you!

---
In reply to: [195849543](https://github.com/dotnet/roslyn/pull/27130#discussion_r195849543) [](ancestors = 195849543)